src/join.rs

```diff
@@ -1,52 +1,224 @@
+pub trait BitAnd {
+    type Value: BitSetLike;
+    fn and(self) -> Self::Value;
 }
 
 /// This needs to be special cased
-impl<A> Join for (A,)
+impl<A> BitAnd for (A,)
     where A: BitSetLike
 {
     type Value = A;
-    fn join(self) -> Self::Value {
+    fn and(self) -> Self::Value {
         self.0
     }
 }
 
-macro_rules! merge_impl {
+macro_rules! bitset_and {
     // use variables to indicate the arity of the tuple
     ($($from:ident),*) => {
-      impl<$($from),*> Join for ($($from),*)
-          where $($from: BitSetLike),*
-      {
-          type Value = BitSetAnd<
-            <<Self as Split>::Left as Join>::Value,
-            <<Self as Split>::Right as Join>::Value
-          >;
-          fn join(self) -> Self::Value {
+        impl<$($from),*> BitAnd for ($($from),*)
+            where $($from: BitSetLike),*
+        {
+            type Value = BitSetAnd<
+                <<Self as Split>::Left as BitAnd>::Value,
+                <<Self as Split>::Right as BitAnd>::Value
+            >;
+            fn and(self) -> Self::Value {
               let (l, r) = self.split();
-              BitSetAnd(l.join(), r.join())
-          }
-      }
-    }
-}
-
-merge_impl!{A, B}
-merge_impl!{A, B, C}
-merge_impl!{A, B, C, D}
-merge_impl!{A, B, C, D, E}
-merge_impl!{A, B, C, D, E, F}
-merge_impl!{A, B, C, D, E, F, G}
-merge_impl!{A, B, C, D, E, F, G, H}
-merge_impl!{A, B, C, D, E, F, G, H, I}
-merge_impl!{A, B, C, D, E, F, G, H, I, J}
-merge_impl!{A, B, C, D, E, F, G, H, I, J, K}
-merge_impl!{A, B, C, D, E, F, G, H, I, J, K, L}
-merge_impl!{A, B, C, D, E, F, G, H, I, J, K, L, M}
-merge_impl!{A, B, C, D, E, F, G, H, I, J, K, L, M, N}
-merge_impl!{A, B, C, D, E, F, G, H, I, J, K, L, M, N, O}
-merge_impl!{A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P}
+              BitSetAnd(l.and(), r.and())
+            }
+        }
+    }
+}
+
+bitset_and!{A, B}
+bitset_and!{A, B, C}
+bitset_and!{A, B, C, D}
+bitset_and!{A, B, C, D, E}
+bitset_and!{A, B, C, D, E, F}
+bitset_and!{A, B, C, D, E, F, G}
+bitset_and!{A, B, C, D, E, F, G, H}
+bitset_and!{A, B, C, D, E, F, G, H, I}
+bitset_and!{A, B, C, D, E, F, G, H, I, J}
+bitset_and!{A, B, C, D, E, F, G, H, I, J, K}
+bitset_and!{A, B, C, D, E, F, G, H, I, J, K, L}
+bitset_and!{A, B, C, D, E, F, G, H, I, J, K, L, M}
+bitset_and!{A, B, C, D, E, F, G, H, I, J, K, L, M, N}
+bitset_and!{A, B, C, D, E, F, G, H, I, J, K, L, M, N, O}
+bitset_and!{A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P}
+
+/// The only purpose of the `Open` trait is to provide a way
+/// to access the `open` or `open_mut` trait in a generic way
+/// This way the fact that the type is immutable or mutable
+/// is not lost when it is used later.
+pub trait Open {
+    type Value;
+    type Mask: BitSetLike;
+    fn open(self) -> (Self::Mask, Self::Value);
+}
+
+impl<'a, T> Open for &'a T
+    where T: Storage
+{
+    type Value = GetRef<'a, <T as Storage>::UnprotectedStorage>;
```

Comment:
Do you actually need `<T as Storage>` if `Storage` is the only constraint of `T`?